Adds PathValidator functionality to YenShortestPathsIterator

SChudakov opened the pull request:
This pull request resolves #827.

- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/Become-a-Contributor>
- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/How-to-make-your-first-%28code%29-contribution>
- [x] I added [unit tests](https://github.com/jgrapht/jgrapht/wiki/Unit-testing)
- [x] I added [documentation](https://github.com/jgrapht/jgrapht/wiki/How-to-write-documentation)
- [x] I followed the [Coding and Style Conventions](https://github.com/jgrapht/jgrapht/wiki/Coding-and-Style-Conventions)
- [x] I **have not** modified `HISTORY.md` or `CONTRIBUTORS.md`
- [x] I ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/YenShortestPathIterator.java:
```diff
@@ -152,18 +163,77 @@ public YenShortestPathIterator(
-            candidatePaths.insert(shortestPath.getWeight(), shortestPath);
-            deviations.put(shortestPath, source);
-            weightsFrequencies.put(shortestPath.getWeight(), 1);
+            V lastValidDeviation = getLastValidDeviation(shortestPath, source);
```
This `if` statement is a heavy computation and should be moved to the `hasNext` or `next` methods. The reason is that the user doesn't expect the constructor to do potentially all the computation.

SChudakov commented:
*Updates summary*

1. Added `KShortestPathsPerformace` benchmark for the implementations of `KShortestPathAlgorithm` interface.
2. Performed test for `YenKShortestPaths` algorithm without `PathValidator` using `networkx` library. Tested correctness of `YenKShortestPaths` with `PathValidator` by comparing output of algorithm with validation to the output without validation. 
3. Fixed bug in `KShortestSimplePaths` as suggested in #892.
4. Performed performance comparison of the changes algorithms. 

Benchmarks were performed on the `Gnp` random graphs with `n = 20`, and `p = 0.1` or `p = 0.2`. For the path validation 10 random edges were selected from the graph and excluded from the shortest paths computation.

Results before changes of the PR:

![before](https://user-images.githubusercontent.com/13834087/78344887-b2521980-75a5-11ea-8e03-2807e30f0b29.png)

Results after changes of this PR:

![after](https://user-images.githubusercontent.com/13834087/78344918-bed67200-75a5-11ea-93c8-7be92aabaf78.png)

The 4th and 6th bars on the last plot are to high to be shown. 
The actual values of those bars are  62784 ms and 113901 ms correspondingly.

SChudakov commented:
After the bug fix for `KShortestSimplePaths` it performance has significantly worsened. 
Comparing to the `YenKShortestPaths` on the random `Gnp` graph with `n = 20` vertices and edge probability `p = 0.2` the performance ratio is::

```
without path validation:        10641 times
with    path validation:        26488 times
```

Since there is such performance gap, I think it is either worth mentioning in the documentation or deprecating the `KShortestSimplePaths` for further removal in later releases.

Toptachamann commented on jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/YenKShortestPathTest.java:
```diff
@@ -106,6 +114,137 @@ public void testLessThanKPaths()
+     * If the overall number of paths between {@code source} and {@code target} is denoted by $n$
+     * and the value of {@code #NUMBER_OF_PATH_TO_ITERATE} is denoted by $m$ then the method
+     * iterates over $p = min\{n, m\}$ such paths and verifies that they are built correctly. The
+     * method uses the {@link KShortestSimplePaths} implementation to verify the order of paths
```
Remove `KShortestSimplePaths`

Toptachamann commented on jgrapht-core/src/test/java/org/jgrapht/perf/shortestpath/KShortestPathsPerformance.java:
```diff
@@ -0,0 +1,182 @@
+    }
+
+    @Benchmark
+    public List<List<GraphPath<Integer, DefaultWeightedEdge>>> testKShortestSimplePaths(RandomGraphState state) {
```
Potentially, we will need to remove this method

Toptachamann commented:
Please, remove the code related to the `KShortestSimplePath` bug fix as it's unrelated to this PR

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/YenShortestPathIterator.java:
```diff
@@ -77,6 +91,12 @@
 
     /**
      * Provides possibility to validate computed paths and exclude invalid ones.
+     * Whenever a candidate path $P$ first deviation vertex $u$ is produces by this
```
'is produces by this algorithm " -> "is produced by this algorithm, "

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/YenShortestPathIterator.java:
```diff
@@ -88,13 +108,18 @@
-     * the path is valid ot not. Invalid paths are kept in the queue, because it is possible
-     * to build a valid path by deviating from an invalid one.
+     * the path is valid ot not. An invalid path is a path which contains an edge which
+     * fails the {@code pathValidator} check.  Invalid paths are kept in the queue, because
```
"  Invalid" -> " Invalid" (2 spaces)

SChudakov commented:
Done.

Toptachamann commented:
Great job, thanks!
